spec/app/ics-interchain-queries/README.md

```diff
@@ -0,0 +1,151 @@
+- `Key`: A Key in the querying chain module is a user specified query identifier for a query so a user can identify various query types. For example, if you were to query stakers/delegators on Gaia (Cosmos Hub) and store it in state. You can set the `key` as `stakers` and thus query via the key later on as needed.
+
+``Note:`` that a query in state can only have one unique key-id representation. For example, if a query with key = stakers & id = 0 is already in state, the query will fail if you try to add another query with key = stakers & id = 0. A query with key = stakers & id = 1 will succeed however (as long as that is not in state already).
+
+### Desired Properties
+
+- Permissionless: A Querying Chain can query a chain and implement cross-chain querying without any approval from a third party or chain governance.
+
+- Minimal Querying Chain Work: A Queried Chain has to do no implementation work or add any module to enable cross chain querying. By utilizing an RPC client on a relayer, this is possible.
+
+- Modular: Adding cross-chain querying should be as easy as implementing a module in your chain.
+
+- Control Queried Data: The Querying Chain should have ultimate control on how to handle queried data. Like querying for a certain query form/type.
+
+- Incentivization: In order to incentivize relayers for participating in interchain queries, a bounty is paid.
+
+## Technical Specification
+
+### General Design 
+
+The Querying Chain starts with the implementation of the Interchain Query Module by adding the module into their chain.
+
+The general flow for interchain queries starts with a Cross Chain Query Request from the Querying Chain which is listened to by relayers. Upon recognition of a cross chain query, relayers utilize a ABCI Query Request to query data from the Queried Chain. Upon success, the relayer submits a `MsgSubmitQueryResult` to the Querying chain with the success flag as 1.
+
+On failure of a query, relayers submit `MsgSubmitQueryResult` with the `success` flag as 0 to the Querying chain. Alternatively on timeout based on the height of the querying chain, the querying chain will submit `SubmitQueryTimeoutResult` with the timeout height specified.
+
+### Data Structures
+
+A CrossChainABCIQueryRequest data type is used to specify the query. Included in this is the `Path` which is the path field of the query i.e: /custom/auth/account. `Key` is the data key to name the query i.e: `pools` or `stakers`. `Id` is the id of the query with each key to id being unique i.e: stakers-0 or stakers-1275 (this example follows key-id format). `TimeoutHeight` specifies the timeout height on the querying chain to timeout the query. `Bounty` is a bounty that is given to the relayer for participating in the query. `ClientId` is used to identify the chain of interest. 
+
+```go
+type CrossChainABCIQueryRequest struct {
```

Comment:
Currently the relayer has complete control over which height on the queried chain to use.

This is not a problem with regular IBC apps, because the IBC protocol ensures that the data queried (commitments, acks) will never change and thus the relayer can query at any height and is guaranteed by the protocol to get the same value.

However, with interchain querying, the keys that are being queried can have different values at different heights.

Thus, a malicious relayer can choose to query a height that has a value that benefits them the most.

In fact, with the current spec a relayer can choose to query at a height before the query was even posted on the querying chain.

Thus, it changes the security model from relayers being purely liveness facilitators to them being able to actively manipulate what data goes on chain.

We should not have a different security model for this feature, since that is bound to cause mistakes. we should remove the ability for relayers to influence the outcome a query by specifying the desired height to be queried in the request itself.

The relayer then **must** return the result from a query at that height, and once again we get back to the original security model where the relayer is just able to affect liveness